File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
    const binaryTreeMaxPathSum = [root.val];

    function traverse(currentNode){
        // If currentNode is null;
        if(currentNode === null) return 0;

        let leftTreeMaxSum = traverse(currentNode.left);
        let rightTreeMaxSum = traverse(currentNode.right);

        // Update the leftTreeMaxSum & rightTreeMaxSum;
        leftTreeMaxSum = Math.max(leftTreeMaxSum, 0);
        rightTreeMaxSum = Math.max(rightTreeMaxSum, 0);

        // Update the binary tree max path sum;
        const newBinaryTreeMaxPathSum = currentNode.val + leftTreeMaxSum + rightTreeMaxSum;
        binaryTreeMaxPathSum[0] = Math.max(binaryTreeMaxPathSum[0], newBinaryTreeMaxPathSum);

        // Return the maximum left or right tree sum;
        return currentNode.val + Math.max(leftTreeMaxSum, rightTreeMaxSum);
    }
    traverse(root);

    return binaryTreeMaxPathSum[0];
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private static int traverse(TreeNode currNode, int[] maxPathSum){
        if(currNode == null){
            return Integer.MIN_VALUE;
        }

        int leftPathSum = traverse(currNode.left, maxPathSum);
        int rightPathSum = traverse(currNode.right, maxPathSum);

        if(leftPathSum < 0){
            leftPathSum = 0;
        }

        if(rightPathSum < 0){
            rightPathSum = 0;
        }

        int newMaxPathSum = leftPathSum + rightPathSum + currNode.val;
        maxPathSum[0] = Math.max(maxPathSum[0], newMaxPathSum);

        return currNode.val + Math.max(leftPathSum, rightPathSum);
    }

    public int maxPathSum(TreeNode root) {
        int[] maxPathSum = {Integer.MIN_VALUE};
        traverse(root, maxPathSum);

        return maxPathSum[0];
    }
}